[4] Get active jobs for a pipeline

d2lam opened the pull request:
- Only lists active jobs (archived = false)
- Lists jobs in the order they appear in workflow, then open PRs.

Solves https://github.com/screwdriver-cd/screwdriver/issues/188

d2lam committed:
Get active jobs for a pipeline

stjohnjohnson commented on lib/pipeline.js:
```diff
@@ -191,6 +192,54 @@ class PipelineModel extends BaseModel {
+
+            // sort them by name. All PR jobs look like "PR-*"
+            prJobs = prJobs.sort((job1, job2) =>
+                parseInt(job1.name.slice(START_INDEX), 10) -
```
Why not make a method `job.PRnum` or something.

d2lam replied:
Good idea. I can do that.

stjohnjohnson commented on lib/pipeline.js:
```diff
@@ -191,6 +192,54 @@ class PipelineModel extends BaseModel {
+                archived: false
+            },
+            paginate: {
+                count: PAGINATE_COUNT, // This limit is set by the matrix restriction
```
This is not correct, there is no 25 limit when you are asking for jobs + PRs.  You need to either ask for 100+ or iterate on this.

d2lam replied:
Sorry this is another copy pasta.

stjohnjohnson commented on lib/pipeline.js:
```diff
@@ -191,6 +192,54 @@ class PipelineModel extends BaseModel {
     /**
+     * Fetch all the active jobs that belong to this pipeline
+     * Sorted by workflow order, followed by PR jobs
+     * @return {[type]} [description]
```
Please fill in

stjohnjohnson commented on lib/pipeline.js:
```diff
@@ -191,6 +192,54 @@ class PipelineModel extends BaseModel {
+     * Sorted by workflow order, followed by PR jobs
+     * @return {[type]} [description]
+     */
+    getActiveJobs() {
```
Why method vs. getter?  Are you passing info into this?

d2lam replied:
Nothing get passed in so yes I can make it a getter.
Or we can make this take in archived = true or false and pagination. 

What would you prefer?

stjohnjohnson replied:
i like option 2, combine that with what we did for `jobs.getBuilds` maybe

d2lam replied:
What do you mean combine with `jobs.getBuilds`?

stjohnjohnson commented:
You should also update line 62 to archive jobs.  And what about where PRs get archived?

d2lam commented:
@stjohnjohnson  This PR is just for getting active jobs. I was going to have the actual archiving in a separate PR.

d2lam committed:
Add a method to get PR number

d2lam commented:
@stjohnjohnson  Addressed your comments:
- Added a function `job.prNum()`
- Allow `getJobs` to take in a `config` param. It get jobs with `archived = false` by default. 
- And idk what you mean by this: "combine that with what we did for jobs.getBuilds maybe"

stjohnjohnson commented on lib/job.js:
```diff
@@ -86,6 +87,18 @@ class Job extends BaseModel {
+     */
+    prNum() {
+        if (!this.isPR()) {
+            return -1;
```
Maybe return `null`?

stjohnjohnson replied:
Can we make this into a getter that is stored on first access?  Otherwise the sort function below will be more expensive then it needs to be.

d2lam replied:
Hmmm I didn't want it to be a getter because not every job is a PR, so it's weird that every job has a get PR number. But it's fine.

d2lam replied:
@stjohnjohnson  I addressed this

stjohnjohnson reviewed (changes requested):
Just one more piece of feedback

stjohnjohnson reviewed (approved):
LGTM

petey commented on lib/job.js:
```diff
@@ -86,6 +87,30 @@ class Job extends BaseModel {
+     * @return {Number}     PR number
+     */
+    get prNum() {
+        delete this.prNum;
```
Why is this so heavy-handed?

The logic in the getter doesn't seem that complicated that it needs to be memoized.

```
return this.isPR() ? parseInt(this.name.slice(START_INDEX), 10) : null;
```

d2lam replied:
@petey addressed.

stjohnjohnson reviewed (approved):
Suuure

petey commented:
👍  
![fetch_me_my_hat_jeeves_](https://cloud.githubusercontent.com/assets/78533/18688581/dc611174-7f39-11e6-8298-7cd9cc130146.gif)
